modules/ui/org.eclipse.fx.ui.preferences/src/main/java/org/eclipse/fx/ui/preferences/page/FieldEditorPreferencePage.java

```diff
@@ -35,91 +35,98 @@
 import javafx.scene.layout.Region;
 
 public abstract class FieldEditorPreferencePage extends BasePreferencePage {
-	
+
 	/**
 	 * CSS Class Name for the grid holding the field editors
 	 */
 	public static final String PAGE_GRID_STYLE = "field-editor-grid"; //$NON-NLS-1$
-	
+
 	/**
 	 * CSS Class Name for the region in which the actions buttons are added
 	 */
 	public static final String PAGE_ACTIONS_STYLE = "field-editor-actions"; //$NON-NLS-1$
-	
+
 	/**
 	 * CSS Class Name for the error message label
 	 */
 	public static final String ERROR_MESSAGE_STYLE = "field-editor-error"; //$NON-NLS-1$
-	
+
 	private BorderPane parent;
 	private GridPane grid;
 	private HBox actions = new HBox();
-	private List<FieldEditor> editors = new ArrayList<>();
+	private List<FieldEditor<?>> editors = new ArrayList<>();
 	private Label errorMessage = new Label();
-	
+
 	public FieldEditorPreferencePage(Memento memento, BorderPane parent) {
 		super(memento);
-		
+
 		this.grid = new GridPane() {
 			@Override
 			public String getUserAgentStylesheet() {
 				return FieldEditorPreferencePage.this.getUserAgentStylesheet();
 			}
 		};
-		
+
+		errorMessage.getStyleClass().add(ERROR_MESSAGE_STYLE);
+
 		this.parent = parent;
 		this.parent.setTop(errorMessage);
 		this.parent.setCenter(grid);
 		this.parent.setBottom(actions);
 		actions.setAlignment(Pos.BASELINE_RIGHT);
-		
+
 		grid.getStyleClass().add(PAGE_GRID_STYLE);
-		actions.getStyleClass().add(PAGE_ACTIONS_STYLE);		
+		actions.getStyleClass().add(PAGE_ACTIONS_STYLE);
 		actions.setMinWidth(Region.USE_PREF_SIZE);
 		grid.setMinWidth(Region.USE_PREF_SIZE);
 		grid.setMinHeight(Region.USE_PREF_SIZE);
 	}
-	
+
 	protected String getUserAgentStylesheet() {
 		return null;
 	}
 
 	@Override
 	public Command<Void> persist() {
-		return Command.createCommand(() -> editors.forEach(FieldEditor::persist));
+		return Command.createCommand(() -> editors.stream()
```

Comment:
I have no strong feelings the only problem this might cause is that if 2 values somehow belong together only one might be persisted. BTW is persisting not an atomic operation?

Reply:
In general I'm not a huge fan of JFace's approach, which forces you to stay on one page if it's invalid. I've already experience cases where e.g. a page is incorrectly implemented and causes errors that just can't be solved, and you have no way to get out (You can close the preferences dialog, but if you reopen it you're stuck on that page again; so you can't modify any other preference page).

We could still prevent "Apply", and make persist() fully atomic (i.e. do nothing if at least one editor is invalid). The only difference with JFace would then be that you're still allowed to switch to a different preference page.

Reply:
As I said I have no strong feelings so let's leave it like it is